pyo3-macros-backend/src/method.rs

```diff
@@ -664,36 +660,41 @@ impl<'a> FnSpec<'a> {
                     Some(cls) => quote!(Some(<#cls as #pyo3_path::PyClass>::NAME)),
                     None => quote!(None),
                 };
-                let arg_names = (0..args.len())
-                    .map(|i| format_ident!("arg_{}", i))
-                    .collect::<Vec<_>>();
                 let future = match self.tp {
-                    FnType::Fn(SelfType::Receiver { mutable: false, .. }) => {
-                        quote! {{
-                            #(let #arg_names = #args;)*
-                            let __guard = unsafe { #pyo3_path::impl_::coroutine::RefGuard::<#cls>::new(&#pyo3_path::impl_::pymethods::BoundRef::ref_from_ptr(py, &_slf))? };
-                            async move { function(&__guard, #(#arg_names),*).await }
-                        }}
-                    }
-                    FnType::Fn(SelfType::Receiver { mutable: true, .. }) => {
+                    // If extracting `self`, we move the `_slf` pointer into the async block. This reduces the lifetime for which the Rust state is considered "borrowed"
+                    // to just when the async block is executing.
+                    //
+                    // TODO: we should do this with all arguments, not just `self`, e.g. https://github.com/PyO3/pyo3/issues/5681
+                    FnType::Fn(SelfType::Receiver { mutable, .. }) => {
+                        let arg_names = (0..args.len())
+                            .map(|i| format_ident!("arg_{}", i))
+                            .collect::<Vec<_>>();
+                        let method = if mutable {
+                            syn::Ident::new("extract_pyclass_ref_mut", Span::call_site())
+                        } else {
+                            syn::Ident::new("extract_pyclass_ref", Span::call_site())
+                        };
```

Comment:
nit: very slightly less verbose:
```suggestion
                        let method = syn::Ident::new(if mutable {
                            "extract_pyclass_ref_mut"
                        } else {
                            "extract_pyclass_ref"
                        },  Span::call_site());
```